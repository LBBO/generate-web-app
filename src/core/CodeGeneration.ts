import { readFile, writeFile } from 'fs/promises'
import { formatWithPrettier } from './FormatCode'

export type JsOrTsImportData = {
  importDefault?: string
  importItems?: string[]
  importAllAs?: string
  sourcePath: string
}

export const addImportToJsOrTsFile = async (
  filePath: string,
  importData: JsOrTsImportData,
): Promise<void> => {
  const fileContent = (await readFile(filePath)).toString()
  const fileContentWithSingleLineInputs = formatWithPrettier(
    fileContent,
    filePath,
    {
      // Forces multiline-inputs to be single-line
      printWidth: Infinity,
    },
  )
  const initialImportsRegex = /^(?:import\W.*\n|\s)*/g
  const initialImports = initialImportsRegex
    .exec(fileContentWithSingleLineInputs)?.[0]
    ?.trim()

  if (initialImports === undefined) {
    throw new Error(`Couldn't find any imports at beginning of line`)
  } else {
    let valuesToImport: string | undefined

    if (importData.importAllAs !== undefined) {
      valuesToImport = `* as ${importData.importAllAs}`
    } else {
      const partsOfValuesToImport: string[] = []

      if (importData.importDefault) {
        partsOfValuesToImport.push(importData.importDefault)
      }

      if (importData.importItems) {
        partsOfValuesToImport.push(
          '{ ' +
            importData.importItems
              .sort((a, b) => {
                if (a < b) {
                  return -1
                } else if (a > b) {
                  return 1
                } else {
                  return 0
                }
              })
              .join(', ') +
            ' }',
        )
      }

      valuesToImport = partsOfValuesToImport.join(', ')
    }

    const newImport = `
import ${valuesToImport ? valuesToImport + ' from' : ''} '${
      importData.sourcePath
    }'
`
    const newFileContent =
      initialImports +
      newImport +
      fileContentWithSingleLineInputs.slice(initialImports.length + 1)

    await writeFile(filePath, formatWithPrettier(newFileContent, filePath))
  }
}
