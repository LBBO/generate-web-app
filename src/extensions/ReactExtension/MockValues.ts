export const defaultIndexTsx = `import React from 'react'
import ReactDOM from 'react-dom'
import './index.css'
import App from './App'
import reportWebVitals from './reportWebVitals'

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root'),
)

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals()
`

export const defaultAppTsx = `import React from 'react'
import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className="App-link"
            href="https://reactjs.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux-toolkit.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className="App-link"
            href="https://react-redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React Redux
          </a>
        </span>
      </header>
    </div>
  )
}

export default App
`

export const appTsxWithAdditionalComponents = `import React from 'react'
import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <PreExistingComponent />
        <AnotherComponent />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className="App-link"
            href="https://reactjs.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux-toolkit.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className="App-link"
            href="https://react-redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React Redux
          </a>
        </span>
      </header>
    </div>
  )
}

export default App
`
