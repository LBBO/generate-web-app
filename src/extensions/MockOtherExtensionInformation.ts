import { Extension } from '../core/Extension'
import { ProjectMetaData } from '../core/userDialog/PerformUserDialog'
import { generateMockPackageManagerStrategy } from '../core/packageManagers/MockPackageManagerStrategy'
import { PackageManagerNames } from '../core/packageManagers/PackageManagerStrategy'
import { generateMockExtension } from './MockExtension'
import { OverrideProperties } from '../types/UtilityTypes'

export const generateMockProjectMetadata = (
  // Override should be ProjectMetaData, but the packageManagerStrategy value
  // will be passed to the mock generator. Also, all values should be optional.
  override?: Partial<
    OverrideProperties<
      ProjectMetaData,
      {
        packageManagerStrategy: Parameters<
          typeof generateMockPackageManagerStrategy
        >[0]
      }
    >
  >,
): ProjectMetaData => ({
  name: override?.name ?? 'mock-project-name',
  packageManagerStrategy: generateMockPackageManagerStrategy(
    override?.packageManagerStrategy,
  ),
  chosenPackageManager:
    override?.chosenPackageManager ?? PackageManagerNames.NPM,
})

type OtherExtensionInformation = Parameters<Extension['run']>[1]
export const generateMockOtherExtensionInformation = (
  override?: Partial<
    OverrideProperties<
      OtherExtensionInformation,
      {
        projectMetadata: Parameters<typeof generateMockProjectMetadata>[0]
      }
    >
  >,
): OtherExtensionInformation => ({
  chosenExtensions: override?.chosenExtensions ?? [
    generateMockExtension({ name: 'Mock extension 1' }),
    generateMockExtension({ name: 'Mock extension 2' }),
    generateMockExtension({ name: 'Mock extension 3' }),
  ],
  projectMetadata: generateMockProjectMetadata(override?.projectMetadata),
})
