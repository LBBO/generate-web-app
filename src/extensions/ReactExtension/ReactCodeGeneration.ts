import type { AdditionalInformationForExtensions } from '../../core/Extension'
import path from 'path'
import { readFile, writeFile } from 'fs/promises'
import { getTypeScriptExtension } from '../Getters'
import { formatWithPrettier } from '../../core/FormatCode'
import { addImportToJsOrTsFile } from '../../core/CodeGeneration'

export const addComponent = async (
  srcFilePath: string,
  componentName: string,
  otherInformation: AdditionalInformationForExtensions,
  exportType: 'default' | 'item' = 'item',
): Promise<void> => {
  const pathToAppComponent = path.join(
    otherInformation.projectMetadata.rootDirectory,
    'src',
    `App.${
      getTypeScriptExtension(otherInformation.chosenExtensions) ? 't' : 'j'
    }sx`,
  )

  await addImportToJsOrTsFile(pathToAppComponent, {
    sourcePath: srcFilePath,
    importItems: exportType === 'item' ? [componentName] : undefined,
    importDefault: exportType === 'default' ? componentName : undefined,
  })

  await readFile(pathToAppComponent)
  await writeFile(pathToAppComponent, '')
}

export const surroundAppWithComponentWithoutImport = async (
  componentOpeningTag: string,
  otherInformation: AdditionalInformationForExtensions,
): Promise<void> => {
  const pathToIndexFile = path.join(
    otherInformation.projectMetadata.rootDirectory,
    'src',
    `index.${
      getTypeScriptExtension(otherInformation.chosenExtensions) ? 't' : 'j'
    }sx`,
  )
  const fileContent = (await readFile(pathToIndexFile)).toString()
  const componentName = componentOpeningTag.match(/<(.*?)\W/)?.[1]

  if (!componentName) {
    throw new Error('Malformed opening tag!')
  }

  const componentClosingTag = `</${componentName}>`
  const newFileContent = fileContent.replace(
    /<App ?\/>/g,
    `${componentOpeningTag}<App />${componentClosingTag}`,
  )
  await writeFile(
    pathToIndexFile,
    formatWithPrettier(newFileContent, pathToIndexFile),
  )
}
