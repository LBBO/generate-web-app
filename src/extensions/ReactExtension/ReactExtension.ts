import type { Extension } from '../../core/Extension'
import { ExtensionCategory } from '../../core/Extension'
import { spawn } from 'child_process'
import { PackageManagerNames } from '../../core/packageManagers/PackageManagerStrategy'
import { ESLintExtension } from '../ESLintExtension'
import { ExtensionIndexes, getTypeScriptExtension } from '../Getters'

export type ReactExtensionOptions = Record<string, never>

export const ReactExtension: Extension = {
  name: 'React',
  index: ExtensionIndexes.REACT,
  description: 'A JavaScript library for building user interfaces.',
  linkToDocumentation: new URL('https://reactjs.org/'),
  // Exclusiveness to Angular is declared in Angular plugin
  // exclusiveTo: [],
  dependsOn: [ESLintExtension],
  category: ExtensionCategory.FRONTEND_FRAMEWORK,

  run: (options, otherInformation) => {
    return new Promise((resolve, reject) => {
      const npxArgs = [
        // Do not ask if the package should be installed
        '--yes',
        '--',
        'create-react-app',
        otherInformation.projectMetadata.name,
      ]

      if (getTypeScriptExtension(otherInformation.chosenExtensions)) {
        npxArgs.push('--template', 'typescript')
      }

      // If yarn is installed, CRA uses it unless --use-npm is set.
      // Thus, if the user chose Yarn as their package manager,
      // no extra action is required.
      if (
        otherInformation.projectMetadata.chosenPackageManager ===
        PackageManagerNames.NPM
      ) {
        npxArgs.push('--use-npm')
      }

      const child_process = spawn('npx', npxArgs, {
        stdio: 'inherit',
        shell: true,
      })

      child_process.on('error', (e) => {
        console.log(process.env.PATH)
        console.error(e)
      })

      child_process.on('close', (statusCode) => {
        console.log(`CRA finished with code ${statusCode}`)

        if (statusCode === 0) {
          resolve()
        } else {
          reject()
        }
      })
    })
  },
}
