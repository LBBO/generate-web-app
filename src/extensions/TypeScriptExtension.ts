import type { Extension } from '../core/Extension'
import { ExtensionCategory } from '../core/Extension'

export type TypeScriptExtensionOptions = {
  enableStrictMode: boolean
}

export const TypeScriptExtension: Extension = {
  name: 'TypeScript',
  description:
    'An open-source language which builds on JavaScript by adding static type definitions.',
  linkToDocumentation: new URL('https://www.typescriptlang.org/'),
  category: ExtensionCategory.JAVASCRIPT_FLAVOR,
  declareCliOptions: (program) => {
    program.option('--ts-strict-mode', 'Install TypeScript in strict mode')
    program.option('--no-ts-strict-mode')
  },
  promptOptions: async (prompt, cliOptions) => {
    const strictModeIsPreEnabled = cliOptions.tsStrictMode as
      | boolean
      | undefined

    if (strictModeIsPreEnabled === undefined) {
      const answers = await prompt<{ typescriptStrictMode: boolean }>([
        {
          name: 'typescriptStrictMode',
          type: 'confirm',
          message:
            'Would you like to enable the TypeScript strict mode? More info: https://www.typescriptlang.org/tsconfig#strict',
          default: true,
        },
      ])
      return {
        enableStrictMode: answers.typescriptStrictMode,
      }
    } else {
      return {
        enableStrictMode: strictModeIsPreEnabled,
      }
    }
  },
  canBeSkipped: (options, otherInformation) => {
    const chosenExtensionNames = otherInformation.chosenExtensions.map(
      (extension) => extension.name,
    )
    return (
      chosenExtensionNames.includes('React') ||
      chosenExtensionNames.includes('Angular')
    )
  },
  run(): Promise<void> {
    return Promise.reject(new Error('TypeScript extension is not implemented'))
  },
}
