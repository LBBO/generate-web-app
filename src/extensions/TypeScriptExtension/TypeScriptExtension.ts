import type { Extension } from '../../core/Extension'
import { ExtensionCategory } from '../../core/Extension'
import {
  ExtensionIndexes,
  getAngularExtension,
  getReactExtension,
} from '../Getters'
import { asyncRunCommand } from '../../core/Utils'
import { readFile, writeFile } from 'fs/promises'
import path from 'path'

export type TypeScriptExtensionOptions = {
  enableStrictMode: boolean
}

export const TypeScriptExtension: Extension = {
  name: 'TypeScript',
  index: ExtensionIndexes.TYPESCRIPT,
  description:
    'An open-source language which builds on JavaScript by adding static type definitions.',
  linkToDocumentation: new URL('https://www.typescriptlang.org/'),
  category: ExtensionCategory.JAVASCRIPT_FLAVOR,
  declareCliOptions: (program) => {
    program.option('--ts-strict-mode', 'Install TypeScript in strict mode')
    program.option('--no-ts-strict-mode')
  },
  promptOptions: async (prompt, cliOptions) => {
    const strictModeIsPreEnabled = cliOptions.tsStrictMode as
      | boolean
      | undefined

    if (strictModeIsPreEnabled === undefined) {
      const answers = await prompt<{ typescriptStrictMode: boolean }>([
        {
          name: 'typescriptStrictMode',
          type: 'confirm',
          message:
            'Would you like to enable the TypeScript strict mode? More info: https://www.typescriptlang.org/tsconfig#strict',
          default: true,
        },
      ])
      return {
        enableStrictMode: answers.typescriptStrictMode,
      }
    } else {
      return {
        enableStrictMode: strictModeIsPreEnabled,
      }
    }
  },
  run: async (options, otherInformation): Promise<void> => {
    if (
      !getAngularExtension(otherInformation.chosenExtensions) &&
      !getReactExtension(otherInformation.chosenExtensions)
    ) {
      // Install Typescript and generate tsconfig.json if no framework was installed.
      // Frameworks do this on their own.
      await otherInformation.projectMetadata.packageManagerStrategy.installDependencies(
        [
          {
            name: 'typescript',
            isDevDependency: true,
          },
        ],
      )

      await asyncRunCommand('npx -p typescript tsc --init', {
        cwd: otherInformation.projectMetadata.rootDirectory,
      })
    }

    const pathToTsConfig = path.join(
      otherInformation.projectMetadata.rootDirectory,
      'tsconfig.json',
    )

    if (!getAngularExtension(otherInformation.chosenExtensions) && options) {
      // Set strict mode only if angular hasn't been chosen as it does it automatically
      const tsConfigContent = (await readFile(pathToTsConfig)).toString()
      const modifiedTsConfigContent = tsConfigContent.replace(
        /"strict": (?:true,( ?)|false,)( ?)/,
        // Ensure the indentation of the comment after "strict": ... stays correct
        `"strict": ${options.enableStrictMode ? 'true,$1$2' : 'false,'}$2`,
      )
      await writeFile(pathToTsConfig, modifiedTsConfigContent)
    }
  },
}
