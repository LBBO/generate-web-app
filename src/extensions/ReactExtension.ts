import { Extension, ExtensionCategory } from '../core/Extension'
import { spawn } from 'child_process'
import { getTypeScriptExtension } from './TypeScriptExtension'
import { PackageManager } from '../core/userDialog/ChoosePackageManager'

export const ReactExtension: Extension = {
  name: 'React',
  description: 'A JavaScript library for building user interfaces.',
  linkToDocumentation: new URL('https://reactjs.org/'),
  // Exclusiveness to Angular is declared in Angular plugin
  exclusiveTo: [],
  category: ExtensionCategory.FRONTEND_FRAMEWORK,

  run: (options, otherInformation) => {
    return new Promise((resolve, reject) => {
      const npxArgs = [
        'create-react-app',
        otherInformation.projectMetadata.name,
      ]

      if (getTypeScriptExtension(otherInformation.chosenExtensions)) {
        npxArgs.push('--template', 'typescript')
      }

      // If yarn is installed, CRA uses it unless --use-npm is set.
      // Thus, if the user chose Yarn as their package manager,
      // no extra action is required.
      if (
        otherInformation.projectMetadata.chosenPackageManager ===
        PackageManager.NPM
      ) {
        npxArgs.push('--use-npm')
      }

      const child_process = spawn('npx', npxArgs, {
        stdio: 'inherit',
      })

      child_process.on('close', (statusCode) => {
        console.log(`CRA finished with code ${statusCode}`)

        if (statusCode === 0) {
          resolve()
        } else {
          reject()
        }
      })
    })
  },
}
