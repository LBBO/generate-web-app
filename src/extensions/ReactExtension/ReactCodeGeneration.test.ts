import fs from 'fs/promises'
import {
  addComponent,
  surroundAppWithComponentWithoutImport,
} from './ReactCodeGeneration'
import { generateMockOtherExtensionInformation } from '../MockOtherExtensionInformation'
import type { AdditionalInformationForExtensions } from '../../core/Extension'
import path from 'path'
import { ReactExtension } from '../ReactExtension'
import { TypeScriptExtension } from '../TypeScriptExtension'
import * as GeneralCodeGeneration from '../../core/CodeGeneration'
import {
  appTsxWithAdditionalComponents,
  defaultAppTsx,
  defaultIndexTsx,
} from './MockValues'

let readFileMock: jest.SpyInstance<ReturnType<typeof fs['readFile']>>
let writeFileMock: jest.SpyInstance<Promise<void>>
let otherExtensionInformation: AdditionalInformationForExtensions

beforeEach(() => {
  readFileMock = jest.spyOn(fs, 'readFile').mockResolvedValue('')
  writeFileMock = jest.spyOn(fs, 'writeFile').mockResolvedValue()
  otherExtensionInformation = generateMockOtherExtensionInformation({
    chosenExtensions: [ReactExtension, TypeScriptExtension],
  })
})

describe('addComponent', () => {
  let addImportToJsOrTsFileMock: jest.SpyInstance

  beforeEach(() => {
    readFileMock.mockResolvedValue(defaultAppTsx)
    addImportToJsOrTsFileMock = jest
      .spyOn(GeneralCodeGeneration, 'addImportToJsOrTsFile')
      .mockResolvedValue(undefined)
  })

  it('should modify App.jsx if typescript is not installed', async () => {
    await addComponent(
      './SomeComponent',
      'SomeComponent',
      generateMockOtherExtensionInformation({
        chosenExtensions: [ReactExtension],
      }),
    )

    const pathToIndexJsx = path.join(
      otherExtensionInformation.projectMetadata.rootDirectory,
      'src',
      'App.jsx',
    )

    expect(readFileMock).toHaveBeenCalledTimes(1)
    expect(readFileMock).toHaveBeenCalledWith(pathToIndexJsx)
    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][0]).toBe(pathToIndexJsx)
  })

  it('should modify App.tsx if typescript is installed', async () => {
    await addComponent(
      './SomeComponent',
      'SomeComponent',
      otherExtensionInformation,
    )

    const pathToIndexTsx = path.join(
      otherExtensionInformation.projectMetadata.rootDirectory,
      'src',
      'App.tsx',
    )

    expect(readFileMock).toHaveBeenCalledTimes(1)
    expect(readFileMock).toHaveBeenCalledWith(pathToIndexTsx)
    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][0]).toBe(pathToIndexTsx)
  })

  it('should add an import for the component', async () => {
    await addComponent(
      './SomeComponent',
      'SomeComponent',
      otherExtensionInformation,
    )

    expect(addImportToJsOrTsFileMock).toHaveBeenCalledTimes(1)
    expect(addImportToJsOrTsFileMock.mock.calls[0][1]).toEqual({
      sourcePath: './SomeComponent',
      importItems: ['SomeComponent'],
    })
  })

  it('should add the component to the end of the App.tsx return statement', async () => {
    await addComponent(
      './SomeComponent',
      'SomeComponent',
      otherExtensionInformation,
    )

    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })

  it('should add the component to the end of the App.tsx return statement', async () => {
    readFileMock.mockResolvedValue(appTsxWithAdditionalComponents)

    await addComponent(
      './NewComponent',
      'NewComponent',
      otherExtensionInformation,
    )

    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })

  it('should support default import', async () => {
    await addComponent(
      './SomeComponent',
      'SomeComponent',
      otherExtensionInformation,
      'default',
    )

    const pathToIndexTsx = path.join(
      otherExtensionInformation.projectMetadata.rootDirectory,
      'src',
      'App.tsx',
    )

    expect(addImportToJsOrTsFileMock).toHaveBeenCalledTimes(1)
    expect(addImportToJsOrTsFileMock).toHaveBeenCalledWith(pathToIndexTsx, {
      sourcePath: './SomeComponent',
      importDefault: 'SomeComponent',
    })
  })
})

describe('surroundAppWithComponentWithoutImport', () => {
  beforeEach(() => {
    readFileMock.mockResolvedValue(defaultIndexTsx)
  })

  it('should modify index.jsx if typescript is not installed', async () => {
    await surroundAppWithComponentWithoutImport(
      '<Switch>',
      generateMockOtherExtensionInformation({
        chosenExtensions: [ReactExtension],
      }),
    )

    const pathToIndexJsx = path.join(
      otherExtensionInformation.projectMetadata.rootDirectory,
      'src',
      'index.jsx',
    )

    expect(readFileMock).toHaveBeenCalledTimes(1)
    expect(readFileMock).toHaveBeenCalledWith(pathToIndexJsx)
    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][0]).toBe(pathToIndexJsx)
  })

  it('should modify index.tsx if typescript is installed', async () => {
    await surroundAppWithComponentWithoutImport(
      '<Switch>',
      otherExtensionInformation,
    )

    const pathToIndexTsx = path.join(
      otherExtensionInformation.projectMetadata.rootDirectory,
      'src',
      'index.tsx',
    )

    expect(readFileMock).toHaveBeenCalledTimes(1)
    expect(readFileMock).toHaveBeenCalledWith(pathToIndexTsx)
    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][0]).toBe(pathToIndexTsx)
  })

  it('should surround the <App /> component with the given provider tag in the index.tsx', async () => {
    await surroundAppWithComponentWithoutImport(
      '<Switch>',
      otherExtensionInformation,
    )

    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })

  it('should accept a component with extra props', async () => {
    await surroundAppWithComponentWithoutImport(
      '<Provider store={store}>',
      otherExtensionInformation,
    )

    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })
})
