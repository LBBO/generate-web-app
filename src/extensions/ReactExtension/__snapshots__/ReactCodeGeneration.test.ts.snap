// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addComponent should add the component to the end of the App.tsx return statement 1`] = `
"import React from 'react'
import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <div className=\\"App\\">
      <header className=\\"App-header\\">
        <img src={logo} className=\\"App-logo\\" alt=\\"logo\\" />
        <SomeComponent />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className=\\"App-link\\"
            href=\\"https://reactjs.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            React
          </a>
          <span>, </span>
          <a
            className=\\"App-link\\"
            href=\\"https://redux.js.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className=\\"App-link\\"
            href=\\"https://redux-toolkit.js.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className=\\"App-link\\"
            href=\\"https://react-redux.js.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            React Redux
          </a>
        </span>
      </header>
    </div>
  )
}

export default App
"
`;

exports[`addComponent should add the component to the end of the App.tsx return statement 2`] = `
"import React from 'react'
import logo from './logo.svg'
import './App.scss'

function App() {
  return (
    <div className=\\"App\\">
      <header className=\\"App-header\\">
        <img src={logo} className=\\"App-logo\\" alt=\\"logo\\" />
        <PreExistingComponent />
        <AnotherComponent />
        <NewComponent />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className=\\"App-link\\"
            href=\\"https://reactjs.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            React
          </a>
          <span>, </span>
          <a
            className=\\"App-link\\"
            href=\\"https://redux.js.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className=\\"App-link\\"
            href=\\"https://redux-toolkit.js.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className=\\"App-link\\"
            href=\\"https://react-redux.js.org/\\"
            target=\\"_blank\\"
            rel=\\"noopener noreferrer\\"
          >
            React Redux
          </a>
        </span>
      </header>
    </div>
  )
}

export default App
"
`;

exports[`surroundAppWithComponentWithoutImport should accept a component with extra props 1`] = `
"import React from 'react'
import ReactDOM from 'react-dom'
import './index.css'
import App from './App'
import reportWebVitals from './reportWebVitals'

ReactDOM.render(
  <React.StrictMode>
    <Provider store={store}>
      <App />
    </Provider>
  </React.StrictMode>,
  document.getElementById('root'),
)

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals()
"
`;

exports[`surroundAppWithComponentWithoutImport should surround the <App /> component with the given provider tag in the index.tsx 1`] = `
"import React from 'react'
import ReactDOM from 'react-dom'
import './index.css'
import App from './App'
import reportWebVitals from './reportWebVitals'

ReactDOM.render(
  <React.StrictMode>
    <Switch>
      <App />
    </Switch>
  </React.StrictMode>,
  document.getElementById('root'),
)

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals()
"
`;
